src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+      __C2 arrays_equals(false, haystackStart, firstNeedleCompare, compLen, retval, rScratch, xmm_tmp3, xmm_tmp4,
+                         false /* char */, knoreg);
+      __ testl(retval, retval);
+      __ jne_b(L_found);
+
+      CLEAR_BIT(mask, index);
+      __ jne(L_innerLoop);
+      __ jmp(L_loopTop);
+
+      // Found exact match.  Compute offset from beginning of haystack
+      __ bind(L_found);
+      __ subq(hsPtrRet, haystack);
+      __ addq(hsPtrRet, index);
+      __ movq(r11, hsPtrRet);
+      __ jmp(L_checkRangeAndReturn);
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //
+    // Small case default:
+    //
+    //  rbx: haystack
+    //  r14: needle
+    //  r13: k - 1
+    //  r12: k
+    //  r10: n - k
+    //  rbp: -1
+    //  rdi: junk
+    //  rsi: n
+    //  rdx: junk
+    //  rcx: junk
+    //  XMM_BYTE_0 - first element of needle broadcast
+    //  XMM_BYTE_K - last element of needle broadcast
+    //
+    //  Haystack always copied to stack, so 32-byte reads OK
+    //  Haystack length < 32
+    //  10 < needle length < 32
+
+    {
+      __ bind(L_smallCaseDefault);
+
+      Label L_innerLoop;
+
+      const Register firstNeedleCompare = rdx;
+      const Register compLen = r9;
+      const Register haystack = rbx;
+      const Register mask = r8;
+      const Register rTmp = rdi;
+      const Register rTmp2 = r13;
+      const Register rTmp3 = rax;
+
+// r14 and r12 will be re-used later in this procedure
+#undef needle
+#define needle r14
+#undef needleLen
+#define needleLen r12
+
+      broadcast_additional_needles(false, 0 /* unknown */, NUMBER_OF_NEEDLE_BYTES_TO_COMPARE, needle, needleLen, rTmp3,
+                                   isUU, isUL, _masm);
+
+      assert(NUMBER_OF_NEEDLE_BYTES_TO_COMPARE <= 4, "Invalid");
+      assert(NUMBER_OF_NEEDLE_BYTES_TO_COMPARE >= 2, "Invalid");
+      __ leaq(firstNeedleCompare, Address(needle, isU ? (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE - 1) * 2
+                                                      : NUMBER_OF_NEEDLE_BYTES_TO_COMPARE - 1));
+      __ leaq(compLen, Address(needleLen,
+                               isU ? -(NUMBER_OF_NEEDLE_BYTES_TO_COMPARE * 2)
+                                   : -(NUMBER_OF_NEEDLE_BYTES_TO_COMPARE)));  // nlen - 2
+                                                                              // elements
+      //  firstNeedleCompare has address of second element of needle
+      //  compLen has length of comparison to do
+
+      compare_haystack_to_needle(false, 0, NUMBER_OF_NEEDLE_BYTES_TO_COMPARE, L_returnRBP, haystack, isU,
+                                 DO_EARLY_BAILOUT, mask, needleLen, rTmp3, _masm);
+
```

Comment:
It is better to pass the broadcasted xmm registers to compare_haystack_to_nedle. Basically pass input, output, and temps to all the methods.